src/test/rgw/rgw_multi/tests.py

```diff
@@ -1479,3 +1479,153 @@ def test_bucket_creation_time():
+        secondary.zone.cluster.admin(cmd)
+
+    # Restart so bucket_checkpoint can actually fetch things from the
+    # secondaries
+    for secondary in zonegroup_conns.rw_zones[1:]:
+        secondary.zone.start()
+
+    zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
+
+def get_obj_names(zone, bucket_name, maxobjs):
+    """
+    Get names of objects in a bucket.
+    """
+    cmd = ['bucket', 'list'] + zone.zone_args()
+    cmd += ['--bucket', bucket_name]
+    cmd += ['--max-entries', str(maxobjs)]
+    objs_json, _ = zone.cluster.admin(cmd, read_only=True)
+    objs = json.loads(objs_json)
+    return [o['name'] for o in objs]
+
+def trash_bucket(zone, bucket_name):
+    """
+    Remove objects through radosgw-admin, zapping bilog to prevent the deletes
+    from replicating.
+    """
+    objs = get_obj_names(zone, bucket_name, 10000)
+    # Delete the objects
+    for obj in objs:
+        cmd = ['object', 'rm'] + zone.zone_args()
+        cmd += ['--bucket', bucket_name]
+        cmd += ['--object', obj]
+        zone.cluster.admin(cmd)
+
+    # Zap the bilog
+    cmd = ['bilog', 'trim'] + zone.zone_args()
+    cmd += ['--bucket', bucket_name]
+    zone.cluster.admin(cmd)
+
+@attr('bucket_reshard')
+def test_zap_init_bucket_sync_run():
+    """
+    Create several generations of objects, trash them, then run bucket sync init
+    and bucket sync run.
+    """
+    zonegroup = realm.master_zonegroup()
+    zonegroup_conns = ZonegroupConns(zonegroup)
+    primary = zonegroup_conns.rw_zones[0]
+
+    bucket = primary.create_bucket(gen_bucket_name())
+    log.debug('created bucket=%s', bucket.name)
+    zonegroup_meta_checkpoint(zonegroup)
+
+    # Write zeroth generation
+    for obj in range(1, 6):
+        k = new_key(primary, bucket.name, f'obj{obj * 11}')
+        k.set_contents_from_string('foo')
+    zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
+
+    # Write several more generations
+    generations = [17, 19, 23, 29, 31, 37]
+    for num_shards in generations:
+        reshard_bucket(primary.zone, bucket.name, num_shards)
+        for obj in range(1, 6):
+            k = new_key(primary, bucket.name, f'obj{obj * num_shards}')
+            k.set_contents_from_string('foo')
+        zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
+
+    # Stop gateways, trash bucket, init, sync, and restart for every secondary
+    for secondary in zonegroup_conns.rw_zones[1:]:
+        secondary.zone.stop()
+
+        trash_bucket(secondary.zone, bucket.name)
+
+        cmd = ['bucket', 'sync', 'init'] + secondary.zone.zone_args()
+        cmd += ['--bucket', bucket.name]
+        cmd += ['--source-zone', primary.name]
+        secondary.zone.cluster.admin(cmd)
+
+        cmd = ['bucket', 'sync', 'run'] + secondary.zone.zone_args()
+        cmd += ['--bucket', bucket.name, '--source-zone', primary.name]
+        secondary.zone.cluster.admin(cmd)
```

Comment:
this tests that the `bucket sync run` command succeeds, and the `zonegroup_bucket_checkpoint()` at the end tests that everything catches up eventually, after all of the rgws are running again

but before we restart our rgws, i think it's also worth checking that `bucket sync run` itself caught up:
```python
secondary.check_bucket_eq(primary, bucket.name)
```

Reply:
`check_bucket_eq` also requires the zone to be up, unfortunately. I don't know if we want to change that, since it's much more thorough than just comparing object names.

I have added a more minimal `bucket_keys_eq` function that goes through radosgw-admin to compare that both buckets have the same object names.